File: rust/sdk/src/contract/builder.rs
use m10_protos::prost::Message;
use m10_protos::sdk::{
    Contract, CreateLedgerTransfer, CreateLedgerTransfers, CreateTransfer, TransferStep,
};
use std::time::{Duration, SystemTime};

pub const DEFAULT_CONTRACT_DURATION: Duration = Duration::from_secs(300);

/// A builder for [`Contract`]
///
/// A [`Contract`] in M10's system is an agreement amoung multiple parties to complete a series of transactions.
/// Each contract contains series of endorsements from each of the parties. Contracts are designed
/// to be executed across ledgers, for scenarios like FX swaps or other multi-currency transactions.
#[derive(Clone, Debug, Default)]
pub struct ContractBuilder {
    transfers: Vec<CreateLedgerTransfer>,
    valid_for: Duration,
}

impl From<CreateLedgerTransfers> for ContractBuilder {
    fn from(transfer_request: CreateLedgerTransfers) -> Self {
        let valid_for = if transfer_request.valid_until == 0 {
            DEFAULT_CONTRACT_DURATION
        } else {
            let valid_until = SystemTime::UNIX_EPOCH
                + std::time::Duration::from_micros(transfer_request.valid_until);
            valid_until
                .duration_since(SystemTime::now())
                .unwrap_or_else(|_| Duration::new(0, 0))
        };
        Self {
            transfers: transfer_request.transfers,
            valid_for,
        }
    }
}

impl From<Vec<CreateLedgerTransfer>> for ContractBuilder {
    fn from(transfers: Vec<CreateLedgerTransfer>) -> Self {
        Self {
            transfers,
            valid_for: DEFAULT_CONTRACT_DURATION,
        }
    }
}

impl ContractBuilder {
    /// Adds a transfer to the contract between two accounts on a particular ledger.
    pub fn transfer(
        mut self,
        ledger_id: &str,
        from_account_id: Vec<u8>,
        to_account_id: Vec<u8>,
        amount: u64,
        memo: Option<&str>,
    ) -> Self {
        let memo = memo.map(m10_protos::metadata::memo);
        self.transfers.push(CreateLedgerTransfer {
            ledger_id: ledger_id.to_string(),
            nonce: fastrand::u64(..),
            transfer: Some(CreateTransfer {
                transfer_steps: vec![TransferStep {
                    from_account_id,
                    to_account_id,
                    amount,
                    metadata: memo.into_iter().collect(),
                }],
            }),
        });
        self
    }

    /// Adds a timeout to the contract
    pub fn valid_for(mut self, duration: Duration) -> Self {
        self.valid_for = duration;
        self
    }

    /// Builds the [`Contract`]
    pub fn build(self) -> Result<Contract, m10_protos::prost::EncodeError> {
        let valid_until = (SystemTime::now() + self.valid_for)
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap()
            .as_micros() as u64;
        let transfer_reqs = CreateLedgerTransfers {
            transfers: self.transfers,
            valid_until,
        };
        let mut transactions = vec![];
        transfer_reqs.encode(&mut transactions)?;
        Ok(Contract {
            transactions,
            endorsements: Vec::new(),
        })
    }
}

File: node/m10-sdk/src/contract/builder.ts
import { Contract, Memo } from "../protobufs/sdk/metadata";
import { type CreateLedgerTransfer,CreateLedgerTransfers } from "../protobufs/sdk/transaction/transaction";
import { convertMemoToAny, TransactionSigner } from "../utils";

import { FinalizedContractExt } from "./contract_ext";



/**
 * A builder for Contract
 *
 * A Contract in M10's system is an agreement amoung multiple parties to complete a series of transactions.
 * Each contract contains series of endorsements from each of the parties. Contracts are designed
 * to be executed across ledgers, for scenarios like FX swaps or other multi-currency transactions.
 */
export class ContractBuilder {

    public transfers: CreateLedgerTransfer[];
    /**
     * Timeout of the contract in milliseconds
     */
    public validFor: number;
    /**
     * Default contract duration in milliseconds
     */
    public static readonly DEFAULT_CONTRACT_DURATION: number = 300_000;


    public constructor() {
        this.transfers = [];
        this.validFor = ContractBuilder.DEFAULT_CONTRACT_DURATION;
    }

    /**
     * Adds a transfer to the contract between two accounts on a particular ledger
     *
     * @returns {ContractBuilder} with modified transfers
     */
    public transfer(
        ledgerId: string,
        fromAccountId: Uint8Array,
        toAccountId: Uint8Array,
        amount: number,
        memo?: string,
    ): ContractBuilder {

        this.transfers.push({
            ledgerId: ledgerId,
            nonce: BigInt(TransactionSigner.generateNonce()),
            transfer: {
                transferSteps: [{
                    fromAccountId: fromAccountId,
                    toAccountId: toAccountId,
                    amount: BigInt(amount),
                    metadata: (
                        memo
                            ? [convertMemoToAny(Memo.create({ plaintext: memo }))]
                            : []
                    ),
                }],
            },
        });

        return this;
    }

    /**
     * Builds the Contract
     */
    public build(): FinalizedContractExt {

        const validUntil = TransactionSigner.getTimestamp(this.validFor);
        const transferReqs = CreateLedgerTransfers.toBinary({ transfers: this.transfers, validUntil: BigInt(validUntil) });

        const contract = Contract.create({
            transactions: transferReqs,
            endorsements: [],
        });

        return new FinalizedContractExt(contract);
    }

    public static default(): ContractBuilder {
        return new ContractBuilder();
    }
}
